pkg/volumes/openstack/volumes.go

```diff
@@ -0,0 +1,356 @@
+	}
+
+	cfg, err := ReadConfig(configFile)
+	if err != nil {
+		return gophercloud.AuthOptions{}, "", false, err
+	}
+
+	return gophercloud.AuthOptions{
+		IdentityEndpoint: cfg.Global.AuthURL,
+		Username:         cfg.Global.Username,
+		UserID:           cfg.Global.UserID,
+		Password:         cfg.Global.Password,
+		TenantID:         cfg.Global.TenantID,
+		TenantName:       cfg.Global.TenantName,
+		DomainID:         cfg.Global.DomainID,
+		DomainName:       cfg.Global.DomainName,
+		AllowReauth:      true,
+	}, cfg.Global.Region, cfg.BlockStorage.IgnoreVolumeAZ, nil
+}
+
+func (stack *OpenstackVolumes) getClients() error {
+	authOption, region, ignoreAZ, err := getCredential()
+	if err != nil {
+		return fmt.Errorf("error building openstack credentials: %v", err)
+	}
+	stack.ignoreAZ = ignoreAZ
+	provider, err := openstack.NewClient(authOption.IdentityEndpoint)
+	if err != nil {
+		return fmt.Errorf("error building openstack storage client: %v", err)
+	}
+	err = openstack.Authenticate(provider, authOption)
+	if err != nil {
+		return fmt.Errorf("error authenticating openstack client: %v", err)
+	}
+
+	cinderClient, err := openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{
+		Type:   "volumev2",
+		Region: region,
+	})
+	if err != nil {
+		return fmt.Errorf("error building storage client: %v", err)
+	}
+	computeClient, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Type:   "compute",
+		Region: region,
+	})
+	if err != nil {
+		return fmt.Errorf("error building compute client: %v", err)
+	}
+	stack.volumeClient = cinderClient
+	stack.computeClient = computeClient
+	return nil
+}
+
+// InternalIP implements Volumes InternalIP
+func (stack *OpenstackVolumes) InternalIP() net.IP {
+	return stack.internalIP
+}
+
+func (stack *OpenstackVolumes) discoverTags() error {
+
+	// Project ID
+	{
+		stack.project = strings.TrimSpace(stack.meta.ProjectID)
+		if stack.project == "" {
+			return fmt.Errorf("project metadata was empty")
+		}
+		glog.Infof("Found project=%q", stack.project)
+	}
+
+	// Instance Name
+	{
+		stack.instanceName = strings.TrimSpace(stack.meta.Name)
+		if stack.instanceName == "" {
+			return fmt.Errorf("instance name metadata was empty")
+		}
+		glog.Infof("Found instanceName=%q", stack.instanceName)
+	}
+
+	// Internal IP & zone
+	{
+
+		var extendedServer struct {
+			servers.Server
+			availabilityzones.ServerAvailabilityZoneExt
+		}
+		err := servers.Get(stack.computeClient, strings.TrimSpace(stack.meta.ServerID)).ExtractInto(&extendedServer)
+		if err != nil {
+			return fmt.Errorf("Failed to retrieve server information from cloud: %v", err)
+		}
+		ip, err := GetServerFixedIP(extendedServer.Addresses, extendedServer.Name)
+		if err != nil {
+			return fmt.Errorf("error querying InternalIP from name: %v", err)
+		}
+		stack.internalIP = net.ParseIP(ip)
+		stack.zone = extendedServer.AvailabilityZone
+		glog.Infof("Found internalIP=%q and zone=%q", stack.internalIP, stack.zone)
```

Comment:
maybe we should update to klog if we have not.

Reply:
Yes - would be great if someone did that, but not a blocker for this PR.

I'll get to it if nobody else beats me to it :-)